### PR TITLE
(#4300) Implement the responsive_favicons module

### LINKED ISSUE #4300: Implement the responsive_favicons module
### Request Description
The [`responsive_favicons`](https://drupal.org/project/responsive_favicons) was found to be a potential fix for #4246. This is not a perfect solution, but may be sufficient.

The major limitations are that the module:
- Does not support custom favicons on path-based domains (e.g. cancer.gov/nano). **This needs to be verified.**
- Does not clear cache when the form is saved.

### Required Tasks

- [ ] Test PR #4285 (or similar) on ACSF to verify support (or lack thereof) for path-based domains.
- [ ] A designer is needed to create a "generic" .svg favicon for platform sites.
  - [ ] From the "generic" .svg file, create the following: (based on [Evil Martians' "How to Favicon"](https://evilmartians.com/chronicles/how-to-favicon-in-2021-six-files-that-fit-most-needs))
    - [ ] 32x32 favicon.ico
    - [ ] 180x180 PNG apple-touch-icon
    - [ ] 192x192 PNG
    - [ ] 512x512 PNG
    - [ ] `manifest.webmanifest` file specifying the 192x192 and 512x512 PNG files)
- [ ] Add the `favicon.ico` at the site root.
- [ ] Add/enable the `responsive_favicons` module.
  - [ ] Give the `site_admin` role permission to access the module's configuration page.
  - [ ] Create a patch to remove the meta tag "ad" for `realfavicongenerator.net`.

We should also create a wiki page documenting the process for generating and installing the 
favicon files.

## Additional Resources

- The `realfavicongenerator.net` site is a potential convenience for generating files, but is not specifically required.
  - Depending on the mechanism used, it may be necessary to edit the manifest to remove "extra" entries).
  - [web site](https://realfavicongenerator.net)
  - [node CLI application](https://realfavicongenerator.net/favicon/node_cli) [GitHub](https://github.com/RealFaviconGenerator/cli-real-favicon)
  - The actual generator appears to be closed source.

### Process for setting the icons on a top-level domain

1. Generate the zip file with icons (see below), specifying that the icons will be put in `/sites/default/files/favicons`.
	1. Be sure to save the generated tags.
2. Login to the CMS site
3. Go to `Configuration` > `User interface` > `Responsive favicons`
4. Set the inputs:
	- Favicions location: Upload zip file
	- Path to favicon files: `/sites/default/files/favicons`
	- Favicon tags: Paste in the generated tags
5. Under "Upload a zip file", click "Choose File" and browse to the generated zip file.
6. Click the "Save configuration" button.
7. Clear the site's Drupal and Varnish caches
8. If applicable, clear the site's Akamai cache

### Process for creating icons with `realfavicongenerator.net` 

URL: https://realfavicongenerator.net/

- Upload the source image ("Select your Favicon image" button)
	- Needs to be at least 70x70, preferably 260x260
- Adjust parameters
  - On the settings tab: margins and background colors as needed for the different platforms. (This will require designer input)
  - Android has manifest options for "Add to Homescreen"
- Under Favicon Generator options
	- Choose "I cannot place favicon files at the root of my web site."
	- Specify the site upload directory (`/sites/default/files/favicons`)
- Generate and download the  favicon zip file.

**Notes:**
- Includes adjustments for quirks of some OS versions, this seems to be based on older versions (e.g. iOS 7, Android Chrome 39)
- [The code in GitHub](https://github.com/RealFaviconGenerator/realfavicongenerator) is just the web site, not the actual generator.  It's at least four years old; there have been no releases in 5 1/2 years.

### CLI interface

GitHub: https://github.com/RealFaviconGenerator/cli-real-favicon

- This is an npm package that's installed as a global tool.
- It's not super robust (e.g. crashes when command line arguments are missing), but does seem to get the job done
- Takes inputs from a JSON file.
	- Unwanted icons can be omitted and will not be generated.
	- Most (but not all) values from the web UI can be found in the sample document (e.g. OS-specific images).
		- The web UI submits a JSON document which loosely maps to the names in the sample document
		- snake_case names map to camelCaseNames
		- the overall webui JSON is loosely the API document's `design` property

### Favicon Output from the Generator for Default Favicon
`<link rel="apple-touch-icon" sizes="180x180" href="/sites/default/files/favicons/apple-touch-icon.png">`
`<link rel="icon" type="image/png" sizes="32x32" href="/sites/default/files/favicons/favicon-32x32.png">`
`<link rel="icon" type="image/png" sizes="16x16" href="/sites/default/files/favicons/favicon-16x16.png">`
`<link rel="manifest" href="/sites/default/files/favicons/site.webmanifest">`
`<link rel="mask-icon" href="/sites/default/files/favicons/safari-pinned-tab.svg" color="#626366">`
`<link rel="shortcut icon" href="/sites/default/files/favicons/favicon.ico">`
`<meta name="msapplication-TileColor" content="#ffffff">`
`<meta name="msapplication-config" content="/sites/default/files/favicons/browserconfig.xml">`
`<meta name="theme-color" content="#ffffff">`

Image

[favicon_package_v0.16.zip](https://github.com/NCIOCPL/cgov-digital-platform/files/15178957/favicon_package_v0.16.zip)

### PR DESCRIPTION
**Pull request by navsunka**
Closes #4300

### CODE REVIEW

File: docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/browserconfig.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/browserconfig.xml:1]: Remove this file.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_core/config/install/responsive_favicons.settings.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+path_type: path
+path: /profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons
+tags:
+  - '<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />'
+  - '<link rel="icon" type="image/png" href="/favicon-32x32.png" sizes="32x32" />'
+  - '<link rel="icon" type="image/png" href="/favicon-16x16.png" sizes="16x16" />'
+  - '<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#c0143c" />'
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_core/config/install/responsive_favicons.settings.yml:7]: Remove these two and replace them with

```html
<link rel="icon" href="/path-to-the-original.svg" type="image/svg+xml">
```

File: docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/site.webmanifest [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+    "name": "",
+    "short_name": "",
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/site.webmanifest:3]: Remove these two lines.

File: docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/site.webmanifest [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        },
+        {
+            "src": "/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/android-chrome-512x512.png",
+            "sizes": "512x512",
+            "type": "image/png"
+        }
+    ],
+    "theme_color": "#ffffff",
+    "background_color": "#ffffff",
+    "display": "standalone"
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/site.webmanifest:18]: Remove these lines.

File: docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/site.webmanifest [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+        {
+            "src": "/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/android-chrome-192x192.png",
+            "sizes": "192x192",
+            "type": "image/png"
+        },
+        {
+            "src": "/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/android-chrome-512x512.png",
+            "sizes": "512x512",
+            "type": "image/png"
+        }
```
Comment (blairlearn) [inline docroot/profiles/custom/cgov_site/themes/custom/cgov/static/images/design-elements/icons/favicons/site.webmanifest:14]: Change the filenames to `favicon-192.png` and `favicon-512.png`. And, of course, make sure the actual filenames match.

**Review by blairlearn — CHANGES_REQUESTED** [review]

**Review by bryanpizzillo — APPROVED** [review]
Looks good.

**Pull request merged**